File: packages/js/src/processor.ts
/**
 * JsonLdEx — Extended JSON-LD Processor
 *
 * Wraps jsonld.js with backward-compatible extensions for AI/ML,
 * security, and validation. All standard JSON-LD 1.1 operations
 * pass through to jsonld.js; extension keywords are processed
 * as an additional layer.
 */

import * as jsonld from 'jsonld';
import { JsonLdExOptions, ResourceLimits, ProvenanceMetadata, ValidationResult, ShapeDefinition } from './types';
import {
  annotate, getConfidence, getProvenance, filterByConfidence,
  aggregateConfidence, AI_ML_CONTEXT,
} from './extensions/ai-ml';
import {
  vectorTermDefinition, validateVector, cosineSimilarity,
  extractVectors, stripVectorsForRdf,
} from './extensions/vector';
import {
  computeIntegrity, verifyIntegrity, integrityContext,
  isContextAllowed, createSecureDocumentLoader,
  enforceResourceLimits, withTimeout, DEFAULT_RESOURCE_LIMITS,
} from './extensions/security';
import { validateNode, validateDocument } from './extensions/validation';
import { JSONLD_EX_NAMESPACE, EXTENSION_KEYWORDS } from './keywords';

export class JsonLdEx {
  private options: JsonLdExOptions;
  private resourceLimits: Required<ResourceLimits>;

  constructor(options: JsonLdExOptions = {}) {
    this.options = options;
    this.resourceLimits = {
      ...DEFAULT_RESOURCE_LIMITS,
      ...options.resourceLimits,
    };
  }

  // ── Core JSON-LD Operations (delegated to jsonld.js) ──────────

  /**
   * Expand a JSON-LD document with extension support.
   * Enforces resource limits and processes extension keywords.
   */
  async expand(doc: any, opts?: any): Promise<any[]> {
    enforceResourceLimits(doc, this.resourceLimits);

    const expandOpts = this.buildProcessorOptions(opts);

    return withTimeout(
      jsonld.expand(doc, expandOpts),
      this.resourceLimits.maxExpansionTime,
      'expansion'
    );
  }

  /**
   * Compact a JSON-LD document.
   */
  async compact(doc: any, ctx: any, opts?: any): Promise<any> {
    enforceResourceLimits(doc, this.resourceLimits);
    const compactOpts = this.buildProcessorOptions(opts);
    return withTimeout(
      jsonld.compact(doc, ctx, compactOpts),
      this.resourceLimits.maxExpansionTime,
      'compaction'
    );
  }

  /**
   * Flatten a JSON-LD document.
   */
  async flatten(doc: any, ctx?: any, opts?: any): Promise<any> {
    enforceResourceLimits(doc, this.resourceLimits);
    const flattenOpts = this.buildProcessorOptions(opts);
    return withTimeout(
      jsonld.flatten(doc, ctx, flattenOpts),
      this.resourceLimits.maxExpansionTime,
      'flattening'
    );
  }

  /**
   * Convert JSON-LD to N-Quads.
   * Strips vector embeddings before conversion (vectors are annotation-only).
   */
  async toRdf(doc: any, opts?: any): Promise<string> {
    enforceResourceLimits(doc, this.resourceLimits);
    const rdfOpts = this.buildProcessorOptions(opts);
    return withTimeout(
      jsonld.toRDF(doc, { ...rdfOpts, format: 'application/n-quads' }) as Promise<string>,
      this.resourceLimits.maxExpansionTime,
      'RDF conversion'
    );
  }

  /**
   * Convert N-Quads to JSON-LD.
   */
  async fromRdf(nquads: string, opts?: any): Promise<any[]> {
    return jsonld.fromRDF(nquads, opts);
  }

  // ── AI/ML Extension Methods ───────────────────────────────────

  /** Create an annotated value with provenance metadata */
  annotate = annotate;

  /** Get the confidence score from a node/value */
  getConfidence(node: any, property?: string): number | undefined {
    if (property) {
      const propValue = node?.[property];
      if (Array.isArray(propValue)) {
        return getConfidence(propValue[0]);
      }
      return getConfidence(propValue);
    }
    return getConfidence(node);
  }

  /** Extract full provenance metadata from a node/value */
  getProvenance(node: any, property?: string): ProvenanceMetadata {
    if (property) {
      const propValue = node?.[property];
      if (Array.isArray(propValue)) {
        return getProvenance(propValue[0]);
      }
      return getProvenance(propValue);
    }
    return getProvenance(node);
  }

  /** Filter graph nodes by minimum confidence on a property */
  filterByConfidence = filterByConfidence;

  /** Aggregate multiple confidence scores */
  aggregateConfidence = aggregateConfidence;

  /** Get the AI/ML context fragment for use in documents */
  getAiMlContext() {
    return AI_ML_CONTEXT;
  }

  // ── Vector Extension Methods ──────────────────────────────────

  /** Create a vector term definition for a context */
  vectorTermDefinition = vectorTermDefinition;

  /** Validate a vector embedding */
  validateVector = validateVector;

  /** Compute cosine similarity between two vectors */
  cosineSimilarity = cosineSimilarity;

  /** Extract all vectors from an expanded node */
  extractVectors = extractVectors;

  /** Strip vectors before RDF conversion */
  stripVectorsForRdf = stripVectorsForRdf;

  // ── Security Extension Methods ────────────────────────────────

  /** Compute integrity hash for a context */
  computeIntegrity = computeIntegrity;

  /** Verify context integrity */
  verifyIntegrity = verifyIntegrity;

  /** Create a context reference with integrity verification */
  integrityContext = integrityContext;

  /** Check if a context URL is allowed */
  isContextAllowed = isContextAllowed;

  // ── Validation Extension Methods ──────────────────────────────

  /** Validate a single node against a shape */
  validateNode = validateNode;

  /** Validate an entire document against shapes */
  validateDocument = validateDocument;

  // ── Utility Methods ───────────────────────────────────────────

  /**
   * Returns the extension namespace IRI.
   */
  get namespace(): string {
    return JSONLD_EX_NAMESPACE;
  }

  /**
   * Returns all extension keywords.
   */
  get extensionKeywords(): readonly string[] {
    return EXTENSION_KEYWORDS;
  }

  // ── Internal ──────────────────────────────────────────────────

  private buildProcessorOptions(opts?: any): any {
    const base: any = { ...opts };

    // Apply secure document loader if allowlist is configured
    if (this.options.contextAllowlist) {
      const defaultLoader = jsonld.documentLoaders.node();
      base.documentLoader = createSecureDocumentLoader(
        defaultLoader,
        this.options.contextAllowlist
      );
    }

    if (this.options.documentLoader) {
      base.documentLoader = this.options.documentLoader;
    }

    if (this.options.base) {
      base.base = this.options.base;
    }

    return base;
  }
}

File: packages/python/src/jsonld_ex/processor.py
"""
JsonLdEx — Extended JSON-LD Processor (Python)

Wraps PyLD with backward-compatible extensions for AI/ML,
security, and validation.
"""

from __future__ import annotations
from typing import Any, Optional

from pyld import jsonld

from jsonld_ex.ai_ml import (
    annotate, get_confidence, get_provenance,
    filter_by_confidence, aggregate_confidence, ProvenanceMetadata,
)
from jsonld_ex.vector import (
    vector_term_definition, validate_vector, cosine_similarity,
    extract_vectors, strip_vectors_for_rdf,
)
from jsonld_ex.security import (
    compute_integrity, verify_integrity, integrity_context,
    is_context_allowed, enforce_resource_limits, DEFAULT_RESOURCE_LIMITS,
)
from jsonld_ex.validation import validate_node, validate_document, ValidationResult


class JsonLdEx:
    """Extended JSON-LD processor wrapping PyLD."""

    def __init__(
        self,
        resource_limits: Optional[dict[str, int]] = None,
        context_allowlist: Optional[dict[str, Any]] = None,
    ):
        self._limits = {**DEFAULT_RESOURCE_LIMITS, **(resource_limits or {})}
        self._allowlist = context_allowlist

    # ── Core Operations ──────────────────────────────────────────

    def expand(self, doc: dict[str, Any], **kwargs: Any) -> list[dict[str, Any]]:
        """Expand a JSON-LD document with resource limit enforcement."""
        enforce_resource_limits(doc, self._limits)
        return jsonld.expand(doc, kwargs)

    def compact(self, doc: dict[str, Any], ctx: Any, **kwargs: Any) -> dict[str, Any]:
        """Compact a JSON-LD document."""
        enforce_resource_limits(doc, self._limits)
        return jsonld.compact(doc, ctx, kwargs)

    def flatten(self, doc: dict[str, Any], ctx: Any = None, **kwargs: Any) -> dict[str, Any]:
        """Flatten a JSON-LD document."""
        enforce_resource_limits(doc, self._limits)
        return jsonld.flatten(doc, ctx, kwargs)

    def to_rdf(self, doc: dict[str, Any], **kwargs: Any) -> str:
        """Convert to N-Quads."""
        enforce_resource_limits(doc, self._limits)
        return jsonld.to_rdf(doc, {**kwargs, "format": "application/n-quads"})

    def from_rdf(self, nquads: str, **kwargs: Any) -> list[dict[str, Any]]:
        """Convert N-Quads to JSON-LD."""
        return jsonld.from_rdf(nquads, kwargs)

    # ── AI/ML Extensions ─────────────────────────────────────────

    annotate = staticmethod(annotate)
    get_confidence = staticmethod(get_confidence)
    get_provenance = staticmethod(get_provenance)
    filter_by_confidence = staticmethod(filter_by_confidence)
    aggregate_confidence = staticmethod(aggregate_confidence)

    # ── Vector Extensions ────────────────────────────────────────

    vector_term_definition = staticmethod(vector_term_definition)
    validate_vector = staticmethod(validate_vector)
    cosine_similarity = staticmethod(cosine_similarity)
    extract_vectors = staticmethod(extract_vectors)
    strip_vectors_for_rdf = staticmethod(strip_vectors_for_rdf)

    # ── Security Extensions ──────────────────────────────────────

    compute_integrity = staticmethod(compute_integrity)
    verify_integrity = staticmethod(verify_integrity)
    integrity_context = staticmethod(integrity_context)
    is_context_allowed = staticmethod(is_context_allowed)

    # ── Validation Extensions ────────────────────────────────────

    validate_node = staticmethod(validate_node)
    validate_document = staticmethod(validate_document)
